src/engine/app.rs

```diff
@@ -19,14 +19,46 @@ use gfx_device;
 use gfx_device::{DisplayConfig, GfxDevice, gfx_types};
 use renderer::{AmbientLight, DirectionalLight, Pipeline, PointLight, target};
 
+/// The engine type, which holds
+/// several useful structs which can be accessed
+/// by engine and game.
+/// It allows for sharing a thread pool,
+/// and it holds the asset manager.
+pub struct Engine {
+    /// The graphics pipeline
+    pub pipe: Pipeline,
+    /// The ecs planner
+    ///
+    /// To get the world, use `world_mut`.
+    pub planner: Planner<()>,
+    /// The asset manager used to submit
+    /// assets to be loaded in parallel
+    pub manager: AssetManager,
+    /// A thread pool which is used
+    /// for asset loading and the ecs.
+    pub pool: Arc<ThreadPool>,
+}
+
+impl Engine {
+    fn new(pipe: Pipeline, planner: Planner<()>,
+           manager: AssetManager, pool: Arc<ThreadPool>) -> Self {
+        Engine {
+            pipe: pipe,
+            planner: planner,
+            manager: manager,
+            pool: pool,
+        }
+    }
+}
+
 /// User-friendly facade for building games. Manages main loop.
 pub struct Application {
+    /// Holds "global" resources per engine.
+    pub engine: Engine,
+
     // Graphics and asset management structs.
     // TODO: Refactor so `pipe` and `gfx_device` are moved into the renderer.
-    assets: AssetManager,
     gfx_device: GfxDevice,
```

Comment:
Maybe move `GfxDevice` inside `Engine`? Then `Application` would just hold an `Engine` struct and main loop related things (time variables and state machine), so `Application` would just handle the main loop and not care about engine internals (they all would live inside `Engine` struct).

Since `Engine` is exposed to `State` methods I think `GfxDevice` should be private (inside `Engine`), because making `GfxDevice` public doesn't add value and the user would be able to shoot himself in the foot in a couple weird ways:
1. call `GfxDevice::render_world` manually from `State::update`.
2. call `GfxDevice::poll_events` manually from `State::update`.

It might make sense to add an `Engine::advance_frame` method, which would be called from `Application::advance_frame` and which would handle all the `Planner::dispatch`/`GfxDevice::render_world` buisness (to make `Engine` more opaque to `Application`), then if we wanted to change how engine internals work we would have to modify `Engine::advance_frame` method, which is cleanly separated from all the main loop logic.